### DISCUSSION TITLE
POC: transient props

### DISCUSSION THREAD

**Discussion by michelengelen** [mattermost/compass-components#87] [Q&A]
# Intro

In order to simplify the structure and lower the barrier for outside contributions the UI platform recently discussed the addition of `styled components` feature [transient props](https://styled-components.com/docs/api#transient-props). We are currently utilizing a different feature ([shouldForwardProp](https://styled-components.com/docs/api#shouldforwardprop)) to prevent props bleeding into the DOM as HTML attributes and potentially messing with the styling, but are aiming to replace this with our [current POC](https://github.com/mattermost/compass-components/pull/81). It removes the need of the extra call to `withConfig` from the `styled` object and the clunky props-blacklisting.

There might also be the possibility where we need a more fine-grained filtering and re-implement a call to `withConfig()` again, but that should only happen where no other solution is sufficient.

Since this is a pretty high impact and we know from recent discussions, that there have been concerns regarding the complexity of `Compass Components` in general we wanted to get the opinion from developers in our organization on this topic.

## What to expect

We will provide 3 possible solutions, each with its own pros and cons, based on a simplified version of the `Button` component for you to discuss:

1. the current approach (calling `withConfig` with the `shouldForwardProp` option)
2. using transient props with props re-assignment (inside the wrapper/public component)
3. using transient props without props re-assignment

> NOTE: every solution will ultimately render the exact same DOM.

## The problem

while we were exploring the possibilities and limitations of `styled components` we found out, that every prop that was named as a valid HTML attribute was forwarded to the DOM (bleeding), even if the attribute was semantically incorrect (e.g. having a `disabled` attribute on a `div`). To prevent that we initially prepended a call to `withConfig` with the `shouldForwardProp` callback config to the styled object.

_this is a pretty simple example with a short blacklist_

```typescript jsx
const ButtonRoot = styled.button.withConfig<PButtonRoot>({
    shouldForwardProp: (property, validator) =>
        Utils.blockProperty(property, ['size', 'width', 'disabled']) && validator(property),
})((props: ThemedStyledProps<PButtonRoot, TTheme>): FlattenSimpleInterpolation => {
    const { theme, size, width, disabled, ...rest } = props;
    // ...
    // this is where the style calculations/definitions take place
    // ...
});
```

we then added a custom validator function to blacklist/block properties from being passed on to the DOM element that is being rendered. Since this can get quite out of hand and the structure became quite clunky and unreadable we discussed if we should implement the usage of `transient props` instead. That is a built in feature of `styled components` which does basically the same, but without the clunky chaining.

_this is how it could look like without the clutter_

```typescript jsx
const ButtonRoot = styled.button(
    (props: ThemedStyledProps<PButtonRoot, TTheme>): FlattenSimpleInterpolation => {
        const { theme, size, width, disabled, ...rest } = props;
        // ...
        // this is where the style calculations/definitions take place
        // ...
    }
);
```

we can potentially reduce the complexity of the structure by a good amount with that. The change comes with a small prize though and that is the reason for this document.

## 1. using `withConfig` and `shouldForwardProp`

This is the approach we are currently taking.

We define the props and set default values on optional props before passing them down to the root component (the `styled component`).

**Button.tsx**

```typescript jsx
const Button: React.FC<PButton> = (props: PButton) => {
    // set default values where props are undefined
    const {
        label,
        size = DEFAULT_BUTTON_SIZE,
        width = DEFAULT_BUTTON_WIDTH,
        disabled = false,
        ...rest
    } = props;

    // combine the props (with potential defaults) with the rest
    const rootProperties = {
        disabled,
        width,
        size,
        ...rest,
    };

    return <ButtonRoot {...rootProperties}>{label}</ButtonRoot>;
};
```

Within our root component we destructure the props and use them to define the styles. Pretty straight forward. To block props from bleeding into the DOM we use the `shouldForwardProp` config/validator. `Utils.blockProperty` is a custom function that blocks the properties defined in the second parameter (the `string[]`). The `validator` function blocks any properties that are not standard HTML-attributes.

**Button.root.tsx**

```typescript jsx
const ButtonRoot = styled.button.withConfig<PButtonRoot>({
    shouldForwardProp: (property, validator) =>
        Utils.blockProperty(property, ['size', 'width', 'disabled']) && validator(property),
})((props: ThemedStyledProps<PButtonRoot, TTheme>): FlattenSimpleInterpolation => {
    const { theme, size, width, disabled, ...rest } = props;
    // ...
    // this is where the style calculations/definitions take place
    // ...
});
```

**Pros**

-   the possibility to fine-grain which prop should be forwarded to HTML
-   we can define standard blacklists that are being used by `Utils.blockProperty`

**Cons**

-   to allow whitelisting we would need to use an additional Utils function
-   it is unclear which prop will get forwarded
-   props might be overlooked easily
-   the blacklist can get pretty long
-   the additional call to `withConfig` is clunky and not very accessible

## 2. using transient props with props re-assignment

Using transient props this way would allow us to remove the clunky `withConfig()` call and write a clearer and more accessible root component. To whitelist props (have them bleed onto the DOM) we could just leave the props name as it is (for example if we would like to have the `role` prop show on the DOM element). This is especially helpful for `aria-*` and `data-*` attributes.

**Button.tsx**

```typescript jsx
const Button: React.FC<PButton> = (props: PButton) => {
    const {
        label,
        size = DEFAULT_BUTTON_SIZE,
        width = DEFAULT_BUTTON_WIDTH,
        disabled = false,
        ...rest
    } = props;

    const rootProperties = {
        $disabled: disabled,
        $width: width,
        $size: size,
        ...rest,
    };

    return <ButtonRoot {...rootProperties}>{label}</ButtonRoot>;
};
```

Inside the root component we would be able to destructure and use the transient props as usual. The internal mechanics of `styled components` prevent the bleed onto the DOM automatically.

**Button.root.tsx**

```typescript jsx
const ButtonRoot = styled.button(
    (props: ThemedStyledProps<PButtonRoot, TTheme>): FlattenSimpleInterpolation => {
        const { theme, $size, $width, $disabled, ...rest } = props;
        // ...
        // this is where the style calculations/definitions take place
        // ...
    }
);
```

> **NOTE**
>
> We could also rename the props in the destructuring to remove the `$` from the prop-name:
>
> ```typescript jsx
> const { theme, $size: size, $width: width, $disabled: disabled, ...rest } = props;
> ```

for this to be working and being typed out we created a new utility type `TransientProperty` that prepends every type property name with a `$` to mark them as a transient property.

**Button.props.ts**

```typescript
type PButton = {
    label: string;
    disabled?: boolean;
    width?: TButtonWidth;
    size?: TButtonSizeToken;
    onClick?: MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>;
    className?: string;
};

// this is where we would need to wrap the props into a utility type that adds the `$` to mark them as transient props
type PButtonRoot = TransientProperty<Pick<PButton, 'disabled' | 'width' | 'size'>>;
```

this is the definition for the `TransientProperty` utility type

```typescript
type TransientProperty<T> = { [P in keyof T & string as `$${P}`]: T[P] };
```

**Pros**

-   easy "black-/whitelisting" of props (to bleed onto the DOM)
-   `Button` API stays the same
-   no need for prop name abstraction (e.g. `role` -> `buttonRole`)
-   clean root component structure
-   we can possibly move some, if not all, of the complex calculations to the wrapper/public component

**Cons**

-   the props type needs to be adjusted
-   adds a layer of abstraction to the props

## 3. using transient props without props re-assignment

This approach is basically the same as in point 2, with the addition, that we rename the props in the `Button` API as well.

**Button.tsx**

```typescript jsx
const Button: React.FC<PButton> = (props: PButton) => {
    const {
        label,
        $size = DEFAULT_BUTTON_SIZE,
        $width = DEFAULT_BUTTON_WIDTH,
        $disabled = false,
        ...rest
    } = props;

    const rootProperties = {
        $disabled,
        $width,
        $size,
        ...rest,
    };

    return <ButtonRoot {...rootProperties}>{label}</ButtonRoot>;
};
```

The root component would look the same as in point 2, but we would need to adjust the props type.

**Button.props.ts**

```typescript
type PButton = {
    label: string;
    $disabled?: boolean;
    $width?: TButtonWidth;
    $size?: TButtonSizeToken;
    onClick?: MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>;
    className?: string;
};
```

As we would change the public API of the component the new way to use/implement it would look like this:

```typescript jsx
<Button
    label={'CLICK ME'}
    $width={'full'}
    $size={'md'}
    $disabled={true}
    variant={'secondary'}
    onClick={handleClick}
/>
```

**Pros**

-   easy "black-/whitelisting" of props (to bleed onto the DOM)
-   no need for prop name abstraction (e.g. `role` -> `buttonRole`)
-   clean root component structure

**Cons**

-   the props type needs to be adjusted
-   adds a layer of abstraction to the props
-   button API changes

**Comment by ghost** [comment]
I am kind of in favour of using transient props with props re-assignment. The approach we tried before that vas viable but it created additional overhead and confusion about props forwarding. it was easy to miss adding the items to the white/black lists and they could potentially grow infinitely. When it comes to personal style I find the transient props more readable. As for the third option, I would prefer to not extend the components API.

**Comment by jwilander** [comment]
How often is bleeding to the DOM an issue and why is it bad? I don't recall having to deal with this issue in the past when writing components. Is this something we're going to need to do for every component that gets added or just a handful? Also, how do other component libraries deal with these issues?

Sorry for all the questions, just trying to wrap my head around the problem some more.

**Comment by michelengelen** [comment]
not a problem at all @jwilander 

1. most of the time the bleeding is not harmful and does not result in styling issues. However there are some cases in which the styling can get messed up, or the attributes have an impact on the functionality of the DOM element. As a side-effect it does not obfuscate the DOM, so it stays readable for debugging (DOM elements can get quite verbose when aria is involved)

2. This is only needed for certain props which are named the same as a HTML attribute that would be passed over to the DOM by default (like `color`, `width`, etc.), so it is definitely not needed for every component.

3. other libraries use this approach or the `shouldForwardProp` approach. as an example: Material UI uses a custom implementation of styled components and adds `shouldForwardProp` whenever needed. The difference is, that their implementation is a bit different and does not chain calls, but rather passes an option object directly into the call to their `styled` function. (as see here: https://github.com/mui-org/material-ui/blob/34eab1594358be7f00d6668f4651c2836008ed86/packages/mui-material/src/Button/Button.js#L59)

Feel free to drop any questions here and I will try to answer all of them.

**Comment by jespino** [comment]
Sorry, I still see this overwhelmingly complex. I can understand why you "tagged" it as "reducing the complexity", but I still see here a degree of complexity that, from my perspective, is not needed or even acceptable. I don't want to keep repeating myself, so I going to reply to your "these are the 3 options, pick one" question.

If I have to choose one, for me the most usable one from the outside is the option 2, because I don't have to think about what is transient and what is not, the option 3, for me, is a bit leaky abstraction. Said that, I think the extra derived type of the option 2 give extra complexity for the people that wants to start contributing, the need to understand what is that thing to understand what is going on.

**Comment by michelengelen** [comment]
Hey @jespino ... first of all: thanks for your reply and the opinion. Highly appreciated.

I just don't know which part of it is too complex from your perspective. If it is the overall re-assignment we can do it in a more traditional way (and potentially reduce some of the complexity here):

```typescript jsx
const Button: React.FC<PButton> = (props: PButton) => {
    const {
        label,
        size = DEFAULT_BUTTON_SIZE,
        width = DEFAULT_BUTTON_WIDTH,
        disabled = false,
        ...rest
    } = props;

    return (
        <ButtonRoot $disabled={disabled} $width={width} $size={size} {...rest}>
            {label}
        </ButtonRoot>
    );
};
```

The utility type to convert properties to transient props is pretty easy to understand in my opinion (just try to do something like that pre-TS4.x :D ... it's horribly complicated).

I agree that we currently do not have inline-documentation on the utility type ... that would definitely help.
Additionally I will add these to the component generation command so that every developer, that follows this workflow, only needs to adjust a small portion of it without caring about how the type works.

Do you think that these measurements would reduce complexity enough?

**Reply by jespino** [reply]
I wasn't talking about the complexity of this changes, this changes, taking into consideration the current architecture are just fine. The problem from my perspective (I already shared it with you), is the current architecture is complicated, and I'm a big fan of "simple better than complex, and complex better than complicated".

I don't want to enter in that discussion here because I think is more an overall approach thing and this discussion is more about this specific change. I just wanted to transmit that every time I see the code of compass-components I feel like it has a huge amount of accidental complexity.

**Reply by michelengelen** [reply]
OK, fair enough ... this is something we might need a different discussion on. I already talked about this with @zefhemel to identify potential improvements that can be made overall, but having an open discussion and getting more feedback in might be a good idea as well. And since we are currently pretty early in the development process we can add those changes in relatively easy. 

Thoughts on this?

**Reply by jespino** [reply]
I'm ok to discuss anything, but I think we are too far from what we think is "simple", more than 450 LOC spread over 6 files to define a button component, sounds like too much to me.

Something more comparable is the Avatar component, that in `widgets` in mattermost are 2 files (scss and tsx files), a total of 137 LOC, in compass components the same thing is 6 files (not including stories.mdx file) and 429 LOC, more than 3 times LOCs for the "same" functionality. To be fair is not exactly the same functionality, but I can achieve the same functionality without reaching that amount of code and files.

It is just too much boilerplate and unneeded "architecture" for very small and simple components. I can understand more architecture is needed for complex components (I don't know if this architecture would help even in that context).

Again, all this is my very own opinion, and I already know that your opinion differs drastically. I want to give this input to let you know my concerns, but as far as you are taking care of the effort is up to you to take this into consideration or not, I'm perfectly fine with compass-components as something that I have to consume, but I wouldn't feel comfortable making changes to that codebase.
